python/cudf_polars/cudf_polars/dsl/ir.py

```diff
@@ -144,20 +147,55 @@ def get_hashable(self) -> Hashable:
         schema_hash = tuple(self.schema.items())
         return (type(self), schema_hash, args)
 
+    # Hacky to avoid type-checking issues, just advertise the
+    # signature. Both mypy and pyright complain if we have an abstract
+    # method that takes arbitrary *args, but the subclasses have
+    # tighter signatures. This complaint is correct because the
+    # subclass is not Liskov-substitutable for the superclass.
+    # However, we know do_evaluate will only be called with the
+    # correct arguments by "construction".
+    do_evaluate: Callable[..., DataFrame]
```

Comment:
Not sure if I like this any better than just using a catch-all signature. For example, mypy seems happy with:

```python
    @classmethod
    def do_evaluate(cls, *args: Any, **kwargs: Any):
        """
        Evaluate the node (given its evaluated children), and return a dataframe.

        Parameters
        ----------
        args
            Non child arguments followed by any evaluated dataframe inputs.
        kwargs
            Key-word arguments. Should be empty!

        Returns
        -------
        DataFrame (on device) representing the evaluation of this plan
        node.

        Raises
        ------
        NotImplementedError
            If we couldn't evaluate things. Ideally this should not occur,
            since the translation phase should pick up things that we
            cannot handle.
        """
        raise NotImplementedError(
            f"Evaluation of plan {type(cls).__name__}"
        )  # pragma: no cover
```

Reply:
If you remove kwargs from the signature (since there are none), it complains though

Reply:
Yes, I agree it's still "off" :)